Timing Logging Enrichment

vicilliar opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Observability feature.

* **What is the current behavior?** (You can also link to an open issue here)
Timing logs are not output at every step of Marqo add_documents and search. They are only returned to the client via `processingTimeMS`.

* **What is the new behavior (if this is a feature change)?**
Breaks down and adds explicit timing logs to relevant parts of the Marqo process. Replaces existing timing devices with `timeit.default_timer`, due to it being faster to instantiate than `datetime.datetime`.

The following parts now have explicit timing logs (by batch):
1. add_documents - pre-processing (total time)
2. add_documents - pre-processing (vectorisation time). Individual vectorisation time is sent as `logger.debug`
3. add_documents - HTTP roundtrip (Marqo <-> Marqo-os)
4. add_documents - Indexing time (from Marqo-os result)
5. search - pre-processing (total time)
6. search - HTTP roundtrip (Marqo <-> Marqo-os)
7. search - post-processing (total time)
8. search - reranking

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No.

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
Yes.

* **Related Python client changes** (link commit/PR here)


* **Related documentation changes** (link commit/PR here)


* **Other information**:


* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

Commit:
prototype timing logging for add docs and search

Commit:
more verbose timing logs

pandu-k commented on src/marqo/tensor_search/tensor_search.py:
```diff
@@ -852,15 +894,35 @@ def _lexical_search(
+    end_search_http_time = timer()
+    total_search_http_time = end_search_http_time - start_search_http_time
+    num_results = len(search_res['hits']['hits'])
+    logger.info(f"search (lexical) roundtrip: took {(total_search_http_time):.3f}s to send search query (roundtrip) to Marqo-os and received {num_results} results.")
```
Can we get the took time from the Marqo-os response too? In the same way we had "Marqo-os index" timing

vicilliar replied:
done

pandu-k commented on src/marqo/tensor_search/tensor_search.py:
```diff
@@ -1003,8 +1067,23 @@ def _vector_text_search(
+    
+    num_responses = len(response["responses"])     # TODO JOSHUA find num_results
+    logger.info(f"search (tensor) roundtrip: took {(total_search_http_time):.3f}s to send search query (roundtrip) to Marqo-os and received {num_responses} responses.")
+
```
Can we get the took time from the Marqo-os response too? In the same way we had "Marqo-os index" timing

pandu-k commented on src/marqo/tensor_search/tensor_search.py:
```diff
@@ -776,12 +811,17 @@ def search(config: Config, index_name: str, text: str, result_count: int = 3, hi
-    search_result["processingTimeMs"] = round(time_taken.total_seconds() * 1000)
+    time_taken = timer() - t0
+    search_result["processingTimeMs"] = round(time_taken * 1000)
     search_result["query"] = text
```
Can we also log the total time for search (along with search-type)?

vicilliar replied:
done

Commit:
more search time logs added

added Marqo-os search time to both lexical and tensor search. added total search time log. updated descriptions of search to say whether multi or normal search is used.

Commit:
timings made for each search in multisearch

pandu-k commented on src/marqo/tensor_search/tensor_search.py:
```diff
@@ -1022,6 +1112,10 @@ def _vector_text_search(
             raise e
 
+    logger.info(f"  search (tensor) Marqo-os processing time: took {(total_os_process_time):.3f}s for Marqo-os to execute all searches.")
+
```
Can we say 

`took ... for Marqo-os to execute the search`

Marqo-os does a parallel `_msearch`, but this is an implementation detail we can abstract away from the Marqo user.

vicilliar replied:
done.

pandu-k reviewed (commented):
left comment

Commit:
merging in mainline

Commit:
fixed merge conflicts with mainline

Commit:
fixed indentation error at verbosely_add_docs

Commit:
reranking log message to only show when reranker is not none
